Finish read-only parity of hex editor

connor4312 opened the pull request:
This finishes the read-only side of the hex editor. Supports keyboard selection, the data inspector, multiple ranges, all that...

![image](https://user-images.githubusercontent.com/2230985/141024152-97227bb2-1919-4adb-a9a8-ff37d314e583.png)

Recapping from what I discussed in standup, the basic model is that there's a scroll container that is sized and just uses native scrolling. Inside that is the sticky-positioned data view (sticky, since setting it to fixed prevents scrolling from propagating). This is a static display of data. Only rows in the view are rendered, and as scroll happens the rows are moved, added, and removed.

I mentioned that I was having some performance issues with Recoil here. I've got this so that it's now _okay_. It's now 'okay'. Selection updates take about 5ms on my macbook, and scrolling takes about 8ms. It's not great but acceptable. If it ends up being problematic we can look at optimizing it; current usage is relatively idiomatic React however.

Note that when testing or evaluating performance, you want to run `npm run watch -- --minify` to create a production build. The performance characteristics of React when built for production are significantly better than in development.

lramos15 commented:
![recording](https://user-images.githubusercontent.com/4544166/141144553-6f212e79-95f9-4339-a683-c65632cf4b69.gif)
Running `npm install && npm run watch -- --minify` leads to a broken experience. The data inspector doesn't load, the rows seem unevenly spaced, and there is some ghosting when moving the focus. Did I miss a step?

connor4312 commented:
Forgot to push my last commit, fixed 😬

lramos15 commented on src/hexDocument.ts:
```diff
@@ -183,15 +176,15 @@ export class HexDocument extends Disposable implements vscode.CustomDocument {
-		await vscode.workspace.fs.writeFile(vscode.Uri.parse(destination.path + ".json"), Buffer.from(JSON.stringify(this.unsavedEdits), "utf-8"));
+	async backup(destination: vscode.Uri): Promise<vscode.CustomDocumentBackup> {
+		const backup = new Backup(destination);
+		await backup.write(this.model.unsavedEdits);
```
So are backups no longer cancellable?

connor4312 replied:
`vscode.workspace.fs.writeFile` is not cancellable 🙁

lramos15 commented on src/hexEditorProvider.ts:
```diff
@@ -171,7 +171,7 @@ export class HexEditorProvider implements vscode.CustomEditorProvider<HexDocumen
 		// Convert the styles and scripts for the webview into webview URIs
 		const scriptUri = webview.asWebviewUri(vscode.Uri.joinPath(this._context.extensionUri, "dist", "editor.js"));
-		const styleUri = webview.asWebviewUri(vscode.Uri.joinPath(this._context.extensionUri, "dist", "hexEdit.css"));
+		const styleUri = webview.asWebviewUri(vscode.Uri.joinPath(this._context.extensionUri, "dist", "editor.css"));
```
Does editor.css get built from the inlined styles? How does this file come to be

connor4312 replied:
Linaria has an esbuild plugin which generates `editor.css` automatically from the styles in JavaScript. Normally I would use styled-components or css modules for this, but this aren't available in esbuild. I'm actually liking Linaria quite a bit, it has no runtime overhead since the styles get compiled to CSS and replaced with just classnames in the source code.

For example this is what editor.css looks like now:

![image](https://user-images.githubusercontent.com/2230985/142040534-7301d53e-a693-412e-b501-9ec4eed35df0.png)

lramos15 commented on src/test/backup.test.ts:
```diff
@@ -0,0 +1,21 @@
+	it("round trips", async () => {
+		const backup = new Backup(await getTempFile());
+		await backup.write(edits);
+		expect(await backup.read()).to.deep.equal(edits);
```
Is there a benefit to chai over mocha's built in `assert.strictDeepEqual`?

connor4312 replied:
For `strictDeepEqual`, not so much, but Chai has additional assertions which makes writing tests easier and/or produces higher quality assertion messages, as well as a plugin API. I don't feel super strongly about it though so I'm fine using just `assert` if you'd prefer.

lramos15 replied:
If you think chai will be beneficial in the future let's keep it. However, do you think we should be using Jest and React's testing library https://testing-library.com/docs/react-testing-library/intro/ so we can test component rendering and those sorts of things?

lramos15 reviewed (changes requested):
I would love to see more comments / explanation around all the special math. I also wasn't great about that and feel like that's the main barrier to the hex editors maintainability. There's a lot of thinking that needs to be done to adjust for all the virtualization and sizing. Why are some things floored vs ceiled? Why are some things -1? Things like this will be forgotten in a couple months

connor4312 commented on media/editor/dataDisplayContext.tsx:
```diff
@@ -0,0 +1,265 @@
+
+/**
+ * Data management context component. Initially we used Recoil for this, but
+ * this ended up introducing performance issues with very many components.
```
Things that use this aren't rerendered automatically, since the instance in the context doesn't change (it's all shallow equality). Instead this class exposes event listeners that are used in hooks (see the end of this file) which allow only the necessary components to be re-rendered. It's actually quite efficient 🙂

connor4312 commented on src/dataInspectorView.ts:
```diff
@@ -7,7 +7,7 @@ import { getNonce } from "./util";
 	public static readonly viewType = "hexEditor.dataInspectorView";
 	private _view?: vscode.WebviewView;
-	private _lastMessage: any;
+	private _lastMessage: unknown;
```
I think so 🙂 We could give the data inspector a similar revamp to use the new `MessageHandler` and associated typings, though I think that's out of scope for this specific PR.

I just prefer `unknown` to `any` in most cases, and here `unknown` reflects the opaqueness of the message more correctly.

connor4312 commented:
Some math has since changed and moved, so I added comments in https://github.com/microsoft/vscode-hexeditor/commit/bcfec6a4d72ee97289b88199178b727007f90434 to the relevant sections.

connor4312 commented on media/editor/dataDisplayContext.tsx:
```diff
@@ -0,0 +1,265 @@
+	/**
+	 * Updates the currently hovered byte.
+	 */
+	public set hoveredByte(byte: FocusedElement | undefined) {
```
Yea, they're pretty similar I don't really like overloading one hook; I considered having a reusable type that the context would contain two of for each of the hovered byte, but they're slightly different enough that I didn't yet. For example when a byte is hovered, both the raw byte and character should appear as hovered, but for focused elements only the single character or byte gets the style.
